# Coding Exercises (Part 5)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate

### Exercise 5: Sorting & Filtering 

In [2]:
# run the cell!
import pandas as pd

In [4]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [6]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

39. __Sort__ the __listings__ DataFrame by the __"Market_Cap"__ column from __high to low__ and __save__ the new order! __Fill in the gaps__!

In [10]:
listings.sort_values(by = "Market_Cap", ascending = False, inplace= True)

__Inspect__ listings! Which Company has the __highest Market Capitalization__ (Name)?

In [12]:
#run the cell!
listings.head()

Symbol Exchange                   Name  Last_Price    Market_Cap  \
11     AAPL   NASDAQ             Apple Inc.      200.62  9.459795e+11   
4109   MSFT   NASDAQ  Microsoft Corporation      120.19  9.221233e+11   
350    AMZN   NASDAQ       Amazon.com, Inc.     1847.33  9.074138e+11   
2767  GOOGL   NASDAQ          Alphabet Inc.     1206.45  8.387076e+11   
2766   GOOG   NASDAQ          Alphabet Inc.     1202.16  8.357253e+11   

      ADR TSO  IPO_Year             Sector  \
11        NaN    1980.0         Technology   
4109      NaN    1986.0         Technology   
350       NaN    1997.0  Consumer Services   
2767      NaN       NaN         Technology   
2766      NaN    2004.0         Technology   

                                             Industry  
11                             Computer Manufacturing  
4109          Computer Software: Prepackaged Software  
350                    Catalog/Specialty Distribution  
2767  Computer Software: Programming, Data Processing  
2766  Computer Software: Programming, Data Processing

The highest Market Capitalization has ... Apple!

40. __Sort__ listings by __IPO_Year__ (ascending) and __Market_Cap__ (descending) and save the new order. 

In [35]:
listings.axes

[Index([ 294, 1439, 6612,   11, 3539, 5511, 1083, 1542, 2439, 1688,
        ...
        6698, 6699, 6700, 6751, 6816, 6818, 6819, 6820, 6830, 6843],
       dtype='int64', length=6852),
 Index(['Symbol', 'Exchange', 'Name', 'Last_Price', 'Market_Cap', 'ADR TSO',
        'IPO_Year', 'Sector', 'Industry'],
       dtype='object')]

In [53]:
listings.sort_values(by = ['IPO_Year' , 'Market_Cap'] , 
                     axis = 'index' , 
                     ascending= [True , False] , 
                     inplace= True)
listings.head(10)

Symbol Exchange                                    Name  Last_Price  \
294    AMAT   NASDAQ                 Applied Materials, Inc.       42.17   
1439   COKE   NASDAQ            Coca-Cola Consolidated, Inc.      294.70   
6612   WDFC   NASDAQ                           WD-40 Company      163.60   
11     AAPL   NASDAQ                              Apple Inc.      200.62   
3539   KLAC   NASDAQ                  KLA-Tencor Corporation      123.34   
5511   SEIC   NASDAQ                 SEI Investments Company       56.06   
1083   CBRL   NASDAQ  Cracker Barrel Old Country Store, Inc.      152.73   
1542   CSII   NASDAQ            Cardiovascular Systems, Inc.       37.99   
2439   FONR   NASDAQ                       Fonar Corporation       20.45   
1688   DAIO   NASDAQ                    Data I/O Corporation        5.40   

        Market_Cap  ADR TSO  IPO_Year                 Sector  \
294   4.003589e+10      NaN    1972.0             Technology   
1439  2.104584e+09      NaN    1972.0  Consumer Non-Durables   
6612  2.257237e+09      NaN    1973.0       Basic Industries   
11    9.459795e+11      NaN    1980.0             Technology   
3539  2.018632e+10      NaN    1980.0          Capital Goods   
5511  8.615867e+09      NaN    1981.0                Finance   
1083  3.671839e+09      NaN    1981.0      Consumer Services   
1542  1.322928e+09      NaN    1981.0            Health Care   
2439  1.326209e+08      NaN    1981.0            Health Care   
1688  4.506536e+07      NaN    1981.0          Capital Goods   

                                               Industry  
294                                      Semiconductors  
1439                Beverages (Production/Distribution)  
6612                                    Major Chemicals  
11                               Computer Manufacturing  
3539                              Electronic Components  
5511                 Investment Bankers/Brokers/Service  
1083                                        Restaurants  
1542                         Medical/Dental Instruments  
2439  Biotechnology: Electromedical & Electrotherape...  
1688                                Electrical Products

__Inspect__ listings! From all Companies that were listed in __1983__, which one has the __highest Market Capitalization__?

In [61]:
#run the cell!
mask_1983 = listings.IPO_Year == 1983.0
listings.sort_values(by = 'Market_Cap', ascending= False).loc[mask_1983]

Symbol Exchange                             Name  Last_Price  \
312    AMGN   NASDAQ                       Amgen Inc.    193.8900   
4730   PAYX   NASDAQ                    Paychex, Inc.     81.1400   
1566   CTAS   NASDAQ               Cintas Corporation    204.9700   
1048   CASY   NASDAQ      Caseys General Stores, Inc.    130.1800   
2919   HCSG   NASDAQ  Healthcare Services Group, Inc.     34.1800   
1027   CAMP   NASDAQ                     CalAmp Corp.     13.0500   
3332   ISRL   NASDAQ                    Isramco, Inc.    119.2900   
6415   VALU   NASDAQ                 Value Line, Inc.     23.2557   
274    ALOT   NASDAQ                  AstroNova, Inc.     24.4800   
5221   RELL   NASDAQ     Richardson Electronics, Ltd.      6.3600   
3587   KTCC   NASDAQ           Key Tronic Corporation      6.0300   

        Market_Cap  ADR TSO  IPO_Year                 Sector  \
312   1.194264e+11      NaN    1983.0            Health Care   
4730  2.916172e+10      NaN    1983.0      Consumer Services   
1566  2.143389e+10      NaN    1983.0  Consumer Non-Durables   
1048  4.768580e+09      NaN    1983.0      Consumer Durables   
2919  2.530550e+09      NaN    1983.0            Health Care   
1027  4.470813e+08      NaN    1983.0             Technology   
3332  3.241882e+08      NaN    1983.0                 Energy   
6415  2.249059e+08      NaN    1983.0                Finance   
274   1.702044e+08      NaN    1983.0             Technology   
5221  8.299495e+07      NaN    1983.0  Consumer Non-Durables   
3587  6.488087e+07      NaN    1983.0             Technology   

                                               Industry  
312   Biotechnology: Biological Products (No Diagnos...  
4730                    Diversified Commercial Services  
1566                                            Apparel  
1048                             Automotive Aftermarket  
2919                        Hospital/Nursing Management  
1027  Radio And Television Broadcasting And Communic...  
3332                               Oil & Gas Production  
6415                                Investment Managers  
274                       Computer peripheral equipment  
5221                              Electronic Components  
3587                      Computer peripheral equipment

... Amgen Inc. has the highest Market Capitalization.

41. Set the column __Symbol__ as new Index of the __listings__ DataFrame! __Save__ the change!

In [63]:
listings.set_index(keys = "Symbol",inplace= True)

42. __Sort__ the __listings__ DataFrame by the new __index__ from __low to high__. __Save__ the new order!

In [81]:
# run the cell!
listings.head()

Symbol Exchange                          Name  Last_Price    Market_Cap  \
0   AMAT   NASDAQ       Applied Materials, Inc.       42.17  4.003589e+10   
1   COKE   NASDAQ  Coca-Cola Consolidated, Inc.      294.70  2.104584e+09   
2   WDFC   NASDAQ                 WD-40 Company      163.60  2.257237e+09   
3   AAPL   NASDAQ                    Apple Inc.      200.62  9.459795e+11   
4   KLAC   NASDAQ        KLA-Tencor Corporation      123.34  2.018632e+10   

   ADR TSO  IPO_Year                 Sector  \
0      NaN    1972.0             Technology   
1      NaN    1972.0  Consumer Non-Durables   
2      NaN    1973.0       Basic Industries   
3      NaN    1980.0             Technology   
4      NaN    1980.0          Capital Goods   

                              Industry  
0                       Semiconductors  
1  Beverages (Production/Distribution)  
2                      Major Chemicals  
3               Computer Manufacturing  
4                Electronic Components

In [141]:
listings.Symbol.copy().sort_values(ascending=True,inplace=True)

In [143]:
listings.set_index('Symbol', inplace = True)

In [147]:
listings.sort_index(ascending= True , inplace= True)
listings.head(10)

Exchange                              Name  Last_Price    Market_Cap  \
Symbol                                                                        
A          NYSE        Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE                 Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                       Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE                AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ     American Airlines Group, Inc.       34.02  1.527687e+10   
AAMC       AMEX  Altisource Asset Management Corp       29.90  4.738157e+07   
AAME     NASDAQ     Atlantic American Corporation        2.48  4.998398e+07   
AAN        NYSE                Aaron&#39;s,  Inc.       53.54  3.628838e+09   
AAOI     NASDAQ     Applied Optoelectronics, Inc.       12.30  2.445561e+08   
AAON     NASDAQ                        AAON, Inc.       44.88  2.336955e+09   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   
AAMC        NaN       NaN           Finance   
AAME        NaN       NaN           Finance   
AAN         NaN       NaN        Technology   
AAOI        NaN    2013.0        Technology   
AAON        NaN       NaN     Capital Goods   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services  
AAMC                                         Real Estate  
AAME                                      Life Insurance  
AAN                      Diversified Commercial Services  
AAOI                                      Semiconductors  
AAON                     Industrial Machinery/Components

In [139]:
# run the cell! (create RangeIndex)
listings.reset_index(inplace = True)

In [145]:
# run the cell!
listings.head(10)

Exchange                                    Name  Last_Price  \
Symbol                                                                
AMAT     NASDAQ                 Applied Materials, Inc.       42.17   
COKE     NASDAQ            Coca-Cola Consolidated, Inc.      294.70   
WDFC     NASDAQ                           WD-40 Company      163.60   
AAPL     NASDAQ                              Apple Inc.      200.62   
KLAC     NASDAQ                  KLA-Tencor Corporation      123.34   
SEIC     NASDAQ                 SEI Investments Company       56.06   
CBRL     NASDAQ  Cracker Barrel Old Country Store, Inc.      152.73   
CSII     NASDAQ            Cardiovascular Systems, Inc.       37.99   
FONR     NASDAQ                       Fonar Corporation       20.45   
DAIO     NASDAQ                    Data I/O Corporation        5.40   

          Market_Cap  ADR TSO  IPO_Year                 Sector  \
Symbol                                                           
AMAT    4.003589e+10      NaN    1972.0             Technology   
COKE    2.104584e+09      NaN    1972.0  Consumer Non-Durables   
WDFC    2.257237e+09      NaN    1973.0       Basic Industries   
AAPL    9.459795e+11      NaN    1980.0             Technology   
KLAC    2.018632e+10      NaN    1980.0          Capital Goods   
SEIC    8.615867e+09      NaN    1981.0                Finance   
CBRL    3.671839e+09      NaN    1981.0      Consumer Services   
CSII    1.322928e+09      NaN    1981.0            Health Care   
FONR    1.326209e+08      NaN    1981.0            Health Care   
DAIO    4.506536e+07      NaN    1981.0          Capital Goods   

                                                 Industry  
Symbol                                                     
AMAT                                       Semiconductors  
COKE                  Beverages (Production/Distribution)  
WDFC                                      Major Chemicals  
AAPL                               Computer Manufacturing  
KLAC                                Electronic Components  
SEIC                   Investment Bankers/Brokers/Service  
CBRL                                          Restaurants  
CSII                           Medical/Dental Instruments  
FONR    Biotechnology: Electromedical & Electrotherape...  
DAIO                                  Electrical Products

43. Check the number of __unique values__ in __all columns__ of the __listings__ DataFrame! __How many__ unique values are in the __"Exchange"__ column?

In [97]:
listings.nunique()

Symbol        6840
Exchange         3
Name          5766
Last_Price    4470
Market_Cap    5835
ADR TSO        140
IPO_Year        42
Sector          12
Industry       135
dtype: int64

There are ... 3 unique values in the "Exchange" column!

44. __Select__ the __10__ listings with the __highest Market_Cap__! What is the Company __Name__ of the __third largest__ listing?

In [99]:
listings.nlargest(n = 10, columns='Market_Cap')

Symbol Exchange                           Name  Last_Price    Market_Cap  \
3      AAPL   NASDAQ                     Apple Inc.      200.62  9.459795e+11   
30     MSFT   NASDAQ          Microsoft Corporation      120.19  9.221233e+11   
443    AMZN   NASDAQ               Amazon.com, Inc.     1847.33  9.074138e+11   
3105  GOOGL   NASDAQ                  Alphabet Inc.     1206.45  8.387076e+11   
793    GOOG   NASDAQ                  Alphabet Inc.     1202.16  8.357253e+11   
1337     FB   NASDAQ                 Facebook, Inc.      177.82  5.074965e+11   
1633   BABA     NYSE  Alibaba Group Holding Limited      186.19  4.826388e+11   
3106    JNJ     NYSE              Johnson & Johnson      135.58  3.610186e+11   
3107      V     NYSE                      Visa Inc.      158.56  3.590806e+11   
3108    XOM     NYSE        Exxon Mobil Corporation       81.56  3.453905e+11   

      ADR TSO  IPO_Year             Sector  \
3         NaN    1980.0         Technology   
30        NaN    1986.0         Technology   
443       NaN    1997.0  Consumer Services   
3105      NaN       NaN         Technology   
793       NaN    2004.0         Technology   
1337      NaN    2012.0         Technology   
1633      NaN    2014.0      Miscellaneous   
3106      NaN       NaN        Health Care   
3107      NaN       NaN      Miscellaneous   
3108      NaN       NaN             Energy   

                                             Industry  
3                              Computer Manufacturing  
30            Computer Software: Prepackaged Software  
443                    Catalog/Specialty Distribution  
3105  Computer Software: Programming, Data Processing  
793   Computer Software: Programming, Data Processing  
1337  Computer Software: Programming, Data Processing  
1633                                Business Services  
3106                            Major Pharmaceuticals  
3107                                Business Services  
3108                         Integrated oil Companies

The third largest company in terms of Market Capitalization is ... Amazon!

45. __Select__ the __15__ listings with the __lowest Prices__ (Penny Stocks)! Which Company has the __fourth lowest__ Stock Price?

In [101]:
listings.nsmallest(n = 15 , columns='Last_Price')

Symbol Exchange                                        Name  Last_Price  \
6373  CYTXZ   NASDAQ                    Cytori Therapeutics Inc.      0.0030   
1630  JASNW   NASDAQ                      Jason Industries, Inc.      0.0040   
2619  VEACW   NASDAQ            Vantage Energy Acquisition Corp.      0.0050   
6837  WHLRW   NASDAQ  Wheeler Real Estate Investment Trust, Inc.      0.0070   
6743  SNOAW   NASDAQ                Sonoma Pharmaceuticals, Inc.      0.0100   
2113  LMFAW   NASDAQ                    LM Funding America, Inc.      0.0115   
6250  APDNW   NASDAQ                    Applied DNA Sciences Inc      0.0151   
2572  AMRWW   NASDAQ                   Alta Mesa Resources, Inc.      0.0207   
6420  EYEGW   NASDAQ               Eyegate Pharmaceuticals, Inc.      0.0213   
6393  DRIOW   NASDAQ                           DarioHealth Corp.      0.0230   
6732  SHIPW   NASDAQ             Seanergy Maritime Holdings Corp      0.0270   
6319  CETXW   NASDAQ                                Cemtrex Inc.      0.0300   
6614  NUROW   NASDAQ                           NeuroMetrix, Inc.      0.0354   
6651  PHIOW   NASDAQ                  Phio Pharmaceuticals Corp.      0.0388   
1897  OXBRW   NASDAQ                Oxbridge Re Holdings Limited      0.0401   

      Market_Cap  ADR TSO  IPO_Year             Sector  \
6373         NaN      NaN       NaN        Health Care   
1630         NaN      NaN    2013.0  Consumer Durables   
2619         NaN      NaN    2017.0            Finance   
6837         NaN      NaN       NaN  Consumer Services   
6743         NaN      NaN       NaN        Health Care   
2113         NaN      NaN    2015.0            Finance   
6250         NaN      NaN       NaN  Consumer Services   
2572         NaN      NaN    2017.0             Energy   
6420         NaN      NaN       NaN        Health Care   
6393         NaN      NaN       NaN        Health Care   
6732         NaN      NaN       NaN     Transportation   
6319         NaN      NaN       NaN      Capital Goods   
6614         NaN      NaN       NaN        Health Care   
6651         NaN      NaN       NaN        Health Care   
1897         NaN      NaN    2014.0            Finance   

                                    Industry  
6373              Medical/Dental Instruments  
1630  Miscellaneous manufacturing industries  
2619                       Business Services  
6837           Real Estate Investment Trusts  
6743                   Major Pharmaceuticals  
2113              Finance: Consumer Services  
6250                 Other Consumer Services  
2572                    Oil & Gas Production  
6420                   Major Pharmaceuticals  
6393              Medical/Dental Instruments  
6732                   Marine Transportation  
6319         Industrial Machinery/Components  
6614              Medical/Dental Instruments  
6651                   Major Pharmaceuticals  
1897              Property-Casualty Insurers

The fourth lowest Stock Price has ... Wheeler Real Estate Investment Trust!

In [ ]:
# run the cell!
listings.head()

In [ ]:
#run the cell!
listings.tail()

46. __Check__ for all elements in the column __"Exchange"__, whether they are __equal to "NASDAQ"__. __Save__ the boolean Series in the variable __mask1__! <br>
__Fill in the gaps__!

In [103]:
mask1 = listings.Exchange == "NASDAQ"
mask1

0        True
1        True
2        True
3        True
4        True
        ...  
6847    False
6848    False
6849    False
6850     True
6851     True
Name: Exchange, Length: 6852, dtype: bool

47. __Check__ for all elements in the column __"Last_Price"__, whether they are __less than 1.0__. __Save__ the boolean Series in the variable __mask2__!

In [105]:
mask2 = listings.Last_Price < 1.0
mask2

0       False
1       False
2       False
3       False
4       False
        ...  
6847    False
6848    False
6849    False
6850    False
6851    False
Name: Last_Price, Length: 6852, dtype: bool

48. __Filter__ the DataFrame __listings__ for all __NASDAQ-listings__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq__ (create a copy)! <br>
__Fill in the gaps__!

In [107]:
nasdaq = listings.loc[mask1].copy()

__How many NASDAQ-listings__ does the __nasdaq__ DataFrame contain?

In [ ]:
# run the cell!
nasdaq.head()

In [109]:
# run the cell!
nasdaq.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3435 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      3435 non-null   object 
 1   Exchange    3435 non-null   object 
 2   Name        3435 non-null   object 
 3   Last_Price  3432 non-null   float64
 4   Market_Cap  3168 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    1599 non-null   float64
 7   Sector      2964 non-null   object 
 8   Industry    2964 non-null   object 
dtypes: float64(4), object(5)
memory usage: 268.4+ KB


There are in total ... 3435 NASDAQ-listings!

49. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) and __save__ the resulting DataFrame in the variable __penny__ (create a copy)!

In [111]:
penny = listings.loc[mask2].copy()

__How many Penny Stocks__ does the __penny__ DataFrame contain?

In [114]:
# run the cell!
penny.head()

Symbol Exchange                             Name  Last_Price  \
64    CYTR   NASDAQ                CytRx Corporation      0.6600   
441   TISA   NASDAQ          Top Image Systems, Ltd.      0.8450   
598   QUIK   NASDAQ           QuickLogic Corporation      0.7410   
605   IGLD   NASDAQ  Internet Gold Golden Lines Ltd.      0.5599   
641   DRRX   NASDAQ               DURECT Corporation      0.6784   

       Market_Cap  ADR TSO  IPO_Year            Sector  \
64   2.220075e+07      NaN    1986.0       Health Care   
441  1.561874e+07      NaN    1996.0        Technology   
598  7.186486e+07      NaN    1999.0        Technology   
605  1.567898e+07      NaN    1999.0  Public Utilities   
641  1.099413e+08      NaN    2000.0       Health Care   

                                              Industry  
64   Biotechnology: Biological Products (No Diagnos...  
441                      Computer peripheral equipment  
598                                     Semiconductors  
605                       Telecommunications Equipment  
641                              Major Pharmaceuticals

In [116]:
# run the cell!
penny.info()

<class 'pandas.core.frame.DataFrame'>
Index: 327 entries, 64 to 6837
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      327 non-null    object 
 1   Exchange    327 non-null    object 
 2   Name        327 non-null    object 
 3   Last_Price  327 non-null    float64
 4   Market_Cap  199 non-null    float64
 5   ADR TSO     6 non-null      float64
 6   IPO_Year    161 non-null    float64
 7   Sector      324 non-null    object 
 8   Industry    324 non-null    object 
dtypes: float64(4), object(5)
memory usage: 25.5+ KB


There are in total ... 327 Penny Stocks!

50. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) that are __NASDAQ-listed__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq_penny__! __Fill in the gaps!__

In [118]:
nasdaq_penny = listings.loc[mask1 & mask2].copy()

__How many listings__ does the __nasdaq_penny__ DataFrame contain?

In [121]:
# run the cell!
nasdaq_penny.head()

Symbol Exchange                             Name  Last_Price  \
64    CYTR   NASDAQ                CytRx Corporation      0.6600   
441   TISA   NASDAQ          Top Image Systems, Ltd.      0.8450   
598   QUIK   NASDAQ           QuickLogic Corporation      0.7410   
605   IGLD   NASDAQ  Internet Gold Golden Lines Ltd.      0.5599   
641   DRRX   NASDAQ               DURECT Corporation      0.6784   

       Market_Cap  ADR TSO  IPO_Year            Sector  \
64   2.220075e+07      NaN    1986.0       Health Care   
441  1.561874e+07      NaN    1996.0        Technology   
598  7.186486e+07      NaN    1999.0        Technology   
605  1.567898e+07      NaN    1999.0  Public Utilities   
641  1.099413e+08      NaN    2000.0       Health Care   

                                              Industry  
64   Biotechnology: Biological Products (No Diagnos...  
441                      Computer peripheral equipment  
598                                     Semiconductors  
605                       Telecommunications Equipment  
641                              Major Pharmaceuticals

In [123]:
# run the cell!
nasdaq_penny.info()

<class 'pandas.core.frame.DataFrame'>
Index: 257 entries, 64 to 6837
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      257 non-null    object 
 1   Exchange    257 non-null    object 
 2   Name        257 non-null    object 
 3   Last_Price  257 non-null    float64
 4   Market_Cap  131 non-null    float64
 5   ADR TSO     6 non-null      float64
 6   IPO_Year    144 non-null    float64
 7   Sector      256 non-null    object 
 8   Industry    256 non-null    object 
dtypes: float64(4), object(5)
memory usage: 20.1+ KB


There are in total ... 257 penny stocks at nasdaq!

# Well Done!

------------------------------

# Hints (Spoiler!)

39. sort_values() method, by parameter

40. pass lists to the by and ascending parameter

41. set_index() method, inplace

42. sort_index() method, inplace

43. nunique() method

44. nlargest() method, parameters n and columns

45. nsmallest() method

46. listings.--- == "---"

47. listings.--- < ---

48. pass mask1 to the loc operator

49. listings.loc[---].---

50. listings.loc[--- & ---]